File: 09-Graph/07-Topological-Sort/topological_sort.c
#include <stdio.h>

#define MAX 100

int main() {
    int n, e;
    int graph[MAX][MAX] = {0};
    int indegree[MAX] = {0};
    int queue[MAX];
    int order[MAX];

    printf("Enter number of vertices: ");
    scanf("%d", &n);

    printf("Enter number of edges: ");
    scanf("%d", &e);

    printf("Enter directed edges (u v):\n");

    for(int i = 0; i < e; i++) {
        int u, v;

        scanf("%d %d", &u, &v);

        graph[u][v] = 1;
        indegree[v]++;
    }

    int front = 0;
    int rear = 0;

    for(int i = 0; i < n; i++) {
        if(indegree[i] == 0) {
            queue[rear++] = i;
        }
    }

    int count = 0;

    while(front < rear) {
        int node = queue[front++];

        order[count++] = node;

        for(int i = 0; i < n; i++) {
            if(graph[node][i]) {
                indegree[i]--;

                if(indegree[i] == 0) {
                    queue[rear++] = i;
                }
            }
        }
    }

    if(count != n) {
        printf("Topological sort is not possible.\n");
        printf("The graph contains a cycle.\n");
    }
    else {
        printf("Topological Order: ");

        for(int i = 0; i < n; i++) {
            printf("%d ", order[i]);
        }

        printf("\n");
    }

    return 0;
}

File: 09-Graph/07-Topological-Sort/topological_sort.py
from collections import deque

n = int(input("Enter number of vertices: "))
e = int(input("Enter number of edges: "))

graph = [[] for _ in range(n)]
indegree = [0] * n

print("Enter directed edges (u v):")

for _ in range(e):
    u, v = map(int, input().split())
    graph[u].append(v)
    indegree[v] += 1

queue = deque()

for i in range(n):
    if indegree[i] == 0:
        queue.append(i)

topological_order = []

while queue:
    node = queue.popleft()
    topological_order.append(node)

    for neighbor in graph[node]:
        indegree[neighbor] -= 1

        if indegree[neighbor] == 0:
            queue.append(neighbor)

if len(topological_order) != n:
    print("Topological sort is not possible.")
    print("The graph contains a cycle.")
else:
    print("Topological Order:", end=" ")

    for node in topological_order:
        print(node, end=" ")

    print()